meter/homematic/connection.go

```diff
@@ -149,3 +145,26 @@ func (c *Connection) GridTotalEnergy() (float64, error) {
 	res, err := c.XmlCmd("getValue", c.MeterChannel, Param{CCUString: "IEC_ENERGY_COUNTER"})
 	return res.Value.CCUFloat, err
 }
+
+// ParseHmError checks on Homematic CCU error codes
+// Refer to page 30 of https://homematic-ip.com/sites/default/files/downloads/HM_XmlRpc_API.pdf
+func ParseHmError(res MethodResponse) error {
+	var faultCode int64
+	var faultString string
+
+	faultCode = 0
+	for i := 0; i < len(res.Fault); i++ {
+		if res.Fault[i].Name == "faultCode" {
+			faultCode = res.Fault[i].Value.CCUInt
+		}
+		if res.Fault[i].Name == "faultString" {
```

Comment:
wäre es nicht einfacher, den String hier zusammen zu bauen und direkt zurück zu geben? Ich rate, aber wäre das ggf. machbar wenn man String und Code direkt gleichwertig im struct ausprägt?

Reply:
Hatte ich versucht. Hab' gestern Nacht keine andere "smartere" Lösung gefunden. In der methodResponse XML ist der faultCode value halt eine Ebene tiefer noch mit einem Data-Type (i4) abgelegt (siehe XML-Beispiel unten).
Hab' extra auch einen Type Test gebaut. Wenn du eine Type-Definition im Kopf hast, die String und Code direkt gleichwertig im struct ausprägt, please share. :-)
```xml
<methodResponse>
  <fault>
    <value>
      <struct>
	    <member>
		  <name>faultCode</name>
		  <value>
		    <i4>-5</i4>
		  </value>
		</member>
		<member>
		  <name>faultString</name>
		  <value>Unknown Parameter for value key: ENERGY_COUNTER</value>
		</member>
      </struct>
	</value>
  </fault>
</methodResponse>
```

Reply:
Ich habe bei meinen Test festgestellt, dass für den gleichen faultcode unterschiedlich Fault Descriptions je nach Device kommen. Deswegen würde ich ihn drin lassen. (der Code ist ja da ..)
Beispiel: Device Fehlercode der Steckdose vs Grid Meter
Charger:
```yaml
thierolm@LAPTOP-KQMMVN3H:~/git-repos/evcc$ ./evcc -c ./evcc_homematictest.yaml -l debug charger
[main  ] INFO 2023/03/15 13:25:47 evcc 0.114.1 (da3ae27d)
[main  ] INFO 2023/03/15 13:25:47 using config file: ./evcc_homematictest.yaml
[db    ] INFO 2023/03/15 13:25:47 using sqlite database: /home/thierolm/.evcc/evcc.db
Power:         Invalid device (-2)
Energy:        Invalid device (-2)
Charge status: Invalid device (-2)
Enabled:       Invalid device (-2)
Features:      []
```
Grid Meter:
```yaml
thierolm@LAPTOP-KQMMVN3H:~/git-repos/evcc$ ./evcc -c ./evcc_homematictest.yaml -l debug meter
[main  ] INFO 2023/03/15 13:43:55 evcc 0.114.1 (da3ae27d)
[main  ] INFO 2023/03/15 13:43:55 using config file: ./evcc_homematictest.yaml
[db    ] INFO 2023/03/15 13:43:55 using sqlite database: /home/thierolm/.evcc/evcc.db
Power:  Unknown instance (-2)
Energy: Unknown instance (-2)
```